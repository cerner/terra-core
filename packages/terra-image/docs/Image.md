| Prop Name | Type | Is Required | Default Value | Description |
|-|-|-|-|-|
| alt| `string`| `required`| `' '`| The text content that specifies an alternative text for an image.|
| height| `string`| optional| | Sets the height of the image.|
| isFluid| `bool`| optional| `false`| Sets the fluid behavior of the image, which is `nonfluid` by default.|
| src| `string`| `required`| | The source for the image which will be displayed.|
| variant| `enum('default','rounded','circle','thumbnail')`| optional| `'default'`| Sets the style of the image from the following values; `default`, `rounded`, `circle`, `thumbnail`.|
| width| `string`| optional| | Sets the width of the image.|