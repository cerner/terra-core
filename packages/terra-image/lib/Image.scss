@import '~terra-mixins';
@import './variables';

.terra-Image {
  display: block;
  vertical-align: middle;
}

.terra-Image--default {
  display: block;
  vertical-align: middle;
}

.terra-Image--rounded {
  border-radius: $terra-image-rounded-border-radius;
}

.terra-Image--circle {
  border-radius: $terra-image-circle-border-radius;
}

.terra-Image--thumbnail {
  @include terra-padding-horizontal($terra-image-thumbnail-padding-left, $terra-image-thumbnail-padding-right);

  background-color: $terra-image-thumbnail-background-color;
  border-color: $terra-image-thumbnail-border-color;
  border-radius: $terra-image-thumbnail-border-radius;
  border-style: $terra-image-thumbnail-border-style;
  border-width: $terra-image-thumbnail-border-width;
  padding-bottom: $terra-image-thumbnail-padding-bottom;
  padding-top: $terra-image-thumbnail-padding-top;
}

.terra-Image--fluid {
  height: auto;
  max-height: 100%;
  max-width: 100%;
}
