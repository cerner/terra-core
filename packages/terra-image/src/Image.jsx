import React from 'react';
import PropTypes from 'prop-types';
import objectFitImages from 'object-fit-images'; // Added polyfill for IE.
import classNames from 'classnames/bind';
import styles from './Image.module.scss';

const cx = classNames.bind(styles);

const ImageVariant = {
  DEFAULT: 'default',
  ROUNDED: 'rounded',
  CIRCLE: 'circle',
  THUMBNAIL: 'thumbnail',
};

const FitTypes = {
  COVER: 'cover',
  SCALEDOWN: 'scale-down',
  FILL: 'fill',
  CONTAIN: 'contain',
  NONE: 'none',
};

const propTypes = {
  /**
   * The source string for the image which will be loaded and displayed.
   */
  src: PropTypes.string.isRequired,
  /**
   * Sets the style of the image from the following values; `default`, `rounded`, `circle`, `thumbnail`.
   */
  variant: PropTypes.oneOf(['default', 'rounded', 'circle', 'thumbnail']),
  /**
   * Sets the fluid behavior of the image, which is `nonfluid` by default.
   */
  isFluid: PropTypes.bool,
  /**
   * The text content that specifies an alternative text for an image.
   */
  alt: PropTypes.string.isRequired,
  /**
   * A React element which will be displayed during loading and in case of src load failure.
   */
  placeholder: PropTypes.element,
  /**
   * Sets the height of the image.
   */
  height: PropTypes.string,
  /**
   * Sets the width of the image.
   */
  width: PropTypes.string,
  /**
   * Function to be executed when the image load is successful.
   */
  onLoad: PropTypes.func,
  /**
   * Function to be executed when the image load errors.
   */
  onError: PropTypes.func,
  /**
  * Sets the `object-fit` style of the image from the following values; `cover`, `contain`, `fill`, `scale-down`, `none`.
  */
  fit: PropTypes.oneOf(['cover', 'scale-down', 'fill', 'contain', 'none']),
};

/* eslint-disable react/default-props-match-prop-types */
const defaultProps = {
  variant: 'default',
  isFluid: false,
  alt: ' ',
  fit: 'fill',
};
/* eslint-enable react/default-props-match-prop-types */

class Image extends React.Component {
  constructor(props) {
    super(props);

    this.ImageRef = React.createRef();

    this.state = {
      isLoading: true,
      isError: false,
      prevPropsSrc: props.src,
    };

    this.handleOnLoad = this.handleOnLoad.bind(this);
    this.handleOnError = this.handleOnError.bind(this);
  }

  componentDidUpdate() {
    objectFitImages(this.ImageRef.current);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.src !== prevState.prevPropsSrc) {
      return {
        isLoading: true,
        isError: false,
        prevPropsSrc: nextProps.src,
      };
    }

    return null;
  }

  handleOnLoad() {
    this.setState({ isLoading: false });
    const { onLoad } = this.props;

    if (onLoad !== undefined) {
      onLoad();
    }
  }

  handleOnError() {
    this.setState({ isLoading: false, isError: true });
    const { onError } = this.props;

    if (onError !== undefined) {
      onError();
    }
  }

  createImage(customProps, imageClasses) {
    const {
      src, alt, height, width,
    } = this.props;
    return (
      <img
        {...customProps}
        src={src}
        alt={alt}
        height={height}
        width={width}
        onLoad={this.handleOnLoad}
        onError={this.handleOnError}
        className={imageClasses}
        ref={this.ImageRef}
      />
    );
  }

  render() {
    const {
      src, variant, isFluid, alt, placeholder, height, width, onLoad, onError, fit, ...customProps
    } = this.props;

    const imageClasses = cx([
      'image',
      fit,
      variant,
      customProps.className,
      { fluid: isFluid },
      { hidden: placeholder && this.state.isLoading },
    ]);

    delete customProps.className;
    if (!this.state.isLoading) {
      objectFitImages(this.ImageRef.current);
    }
    if (placeholder) {
      if (this.state.isLoading) {
        return (
          <>
            {this.createImage(customProps, imageClasses)}
            {placeholder}
          </>
        );
      }

      return this.state.isError ? placeholder : this.createImage(customProps, imageClasses);
    }

    return this.createImage(customProps, imageClasses);
  }
}

Image.propTypes = propTypes;
Image.defaultProps = defaultProps;

export default Image;
export { ImageVariant, FitTypes };
