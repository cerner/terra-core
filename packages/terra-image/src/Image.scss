/* stylelint-disable selector-class-pattern  */
:local {
  .image {
    display: block;
    vertical-align: middle;
  }

  .default {
    display: block;
    vertical-align: middle;
  }

  .rounded {
    border-radius: var(--terra-image-rounded-border-radius, 0.3rem);
  }

  .circle {
    border-radius: var(--terra-image-circle-border-radius, 50%);
  }

  .thumbnail {
    background-color: var(--terra-image-thumbnail-background-color, #fff);
    border-color: var(--terra-image-thumbnail-border-color, #ddd);
    border-radius: var(--terra-image-thumbnail-border-radius, 0.25em);
    border-style: var(--terra-image-thumbnail-border-style, solid);
    border-width: var(--terra-image-thumbnail-border-width, 1px);
    padding-bottom: var(--terra-image-thumbnail-padding-bottom, 0.25rem);
    padding-left: var(--terra-image-thumbnail-padding-left, 0.25rem);
    padding-right: var(--terra-image-thumbnail-padding-right, 0.25rem);
    padding-top: var(--terra-image-thumbnail-padding-top, 0.25rem);
  }

  .fluid {
    height: auto;
    max-height: 100%;
    max-width: 100%;
  }
}
/* stylelint-enable */
