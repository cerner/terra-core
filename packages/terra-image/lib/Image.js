'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

require('./Image.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  /**
   * The source for the image which will be displayed.
   */
  src: _react.PropTypes.string.isRequired,
  /**
   * Sets the style of the image from the following values; `default`, `rounded`, `circle`, `thumbnail`.
   */
  variant: _react.PropTypes.oneOf(['default', 'rounded', 'circle', 'thumbnail']),
  /**
   * Sets the fluid behavior of the image, which is `nonfluid` by default.
   */
  isFluid: _react.PropTypes.bool,
  /**
   * The text content that specifies an alternative text for an image.
   */
  alt: _react.PropTypes.string.isRequired,
  /**
   * Sets the height of the image.
   */
  height: _react.PropTypes.number,
  /**
   * Sets the width of the image.
   */
  width: _react.PropTypes.number
};

var defaultProps = {
  variant: 'default',
  isFluid: false,
  alt: ' '
};

var Image = function Image(_ref) {
  var src = _ref.src,
      variant = _ref.variant,
      isFluid = _ref.isFluid,
      alt = _ref.alt,
      height = _ref.height,
      width = _ref.width,
      customProps = _objectWithoutProperties(_ref, ['src', 'variant', 'isFluid', 'alt', 'height', 'width']);

  var classes = (0, _classnames2.default)(['terra-Image', _defineProperty({}, 'terra-Image--' + variant, variant), { 'terra-Image--fluid': isFluid }, customProps.className]);

  return _react2.default.createElement('img', _extends({
    src: src,
    alt: alt,
    height: height,
    width: width
  }, customProps, {
    className: classes
  }));
};

Image.propTypes = propTypes;
Image.defaultProps = defaultProps;

exports.default = Image;