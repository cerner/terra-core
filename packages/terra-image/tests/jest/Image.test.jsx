import React from 'react';
import Image from '../../src/Image';

const imagecontainerStyle1 = {
  backgroundColor: '#ffe4c4',
  width: 200,
  height: 200,
};

const image = <Image src="" alt="circle" variant="circle" isFluid />;

/* Default component */
it('should render a default component', () => {
  const wrapper = shallow(<Image src="" alt=" " />);
  expect(wrapper.instance().props.alt).toEqual(' ');
});

/* Default component with a height and width */
it('should render a default image component with a specified height and width', () => {
  const wrapper = shallow(<Image src="" height="200" width="300" alt="default image" />);
  expect(wrapper.instance().props.alt).toEqual('default image');
  expect(wrapper.instance().props.height).toEqual('200');
  expect(wrapper.instance().props.width).toEqual('300');
});

/* Image with a rounded variant */
it('should render a rounded image component', () => {
  const wrapper = shallow(<Image src="" variant="rounded" alt="rounded image" />);
  expect(wrapper.instance().props.alt).toEqual('rounded image');
});

/* Image with a circle variant */
it('should render a circle image component', () => {
  const wrapper = shallow(<Image src="" variant="circle" alt="circle image" />);
  expect(wrapper.instance().props.alt).toEqual('circle image');
});

/* Image with a thumbnail variant */
it('should render a thumbnail image component', () => {
  const wrapper = shallow(<Image src="" variant="thumbnail" alt="thumbnail image" />);
  expect(wrapper.instance().props.alt).toEqual('thumbnail image');
});

/* Image with fluid behavior */
it('should render a thumbail image component with fluid behavior', () => {
  const wrapper = render(<Image src="" variant="thumbnail" alt="thumbnail image" isFluid />);
  expect(wrapper).toMatchSnapshot();
});

/* Image without fluid behavior */
it('should render a thumbail image component with non-fluid behavior', () => {
  const wrapper = render(<Image src="" variant="thumbnail" alt="thumbnail image" />);
  expect(wrapper).toMatchSnapshot();
});

/* Image with fluid behavior having a parent container */
it('should render a circle image component with fluid behavior', () => {
  /* eslint-disable react/forbid-dom-props */
  const wrapper = shallow(<div style={imagecontainerStyle1}>{image}</div>);
  expect(wrapper).toMatchSnapshot();
  /* eslint-enable react/forbid-dom-props */
});

/* Image with onLoad function */
it('should render an image with the supplied onLoad function', () => {
  const onLoadFunc = () => {};
  const wrapper = shallow(<Image src="" alt=" " onLoad={onLoadFunc} />);
  expect(wrapper.instance().props.onLoad).toBe(onLoadFunc);
});

/* Image with onError function */
it('should render an image with the supplied onError function', () => {
  const onErrorFunc = () => {};
  const wrapper = shallow(<Image src="" alt=" " onError={onErrorFunc} />);
  expect(wrapper.instance().props.onError).toBe(onErrorFunc);
});

/* Placeholder image is displayed while the src image is loading */
it('should render a hidden src image and a visible placeholder image', () => {
  const wrapper = shallow(<Image src="profile.jpg" alt=" " height="75" width="75" placeholder={<div>placeholder text</div>} />);
  expect(wrapper).toMatchSnapshot();
});

/* Src image successfully loaded */
it('should render the src image', () => {
  const wrapper = shallow(<Image src="profile.jpg" alt="profile" height="75" width="75" placeholder={<div>placeholder text</div>} />);
  wrapper.setState({ isLoading: false });
  expect(wrapper).toMatchSnapshot();
});

/* Src image failed to load */
it('should render the placeholder image', () => {
  const wrapper = shallow(<Image src="profile.jpg" alt="avatar" height="75" width="75" placeholder={<div>placeholder text</div>} />);
  wrapper.setState({ isLoading: false, isError: true });
  expect(wrapper).toMatchSnapshot();
});
