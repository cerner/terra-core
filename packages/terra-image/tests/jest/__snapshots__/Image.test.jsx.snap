// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly applies the theme context className 1`] = `
<ThemeContextProvider
  theme={
    Object {
      "className": "orion-fusion-theme",
    }
  }
>
  <Image
    alt=" "
    fit="fill"
    isFluid={false}
    src=""
    variant="default"
  >
    <img
      alt=" "
      className="image fill default orion-fusion-theme"
      onError={[Function]}
      onLoad={[Function]}
      src=""
    />
  </Image>
</ThemeContextProvider>
`;

exports[`should render a circle image component with fluid behavior 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#ffe4c4",
      "height": 200,
      "width": 200,
    }
  }
>
  <Image
    alt="circle"
    fit="fill"
    isFluid={true}
    src=""
    variant="circle"
  />
</div>
`;

exports[`should render a hidden src image and a visible placeholder image 1`] = `
<Fragment>
  <img
    alt=" "
    className="image fill default hidden"
    height="75"
    onError={[Function]}
    onLoad={[Function]}
    src="profile.jpg"
    width="75"
  />
  <div>
    placeholder text
  </div>
</Fragment>
`;

exports[`should render a thumbail image component with fluid behavior 1`] = `
<img
  alt="thumbnail image"
  class="image fill thumbnail fluid"
  src=""
/>
`;

exports[`should render a thumbail image component with non-fluid behavior 1`] = `
<img
  alt="thumbnail image"
  class="image fill thumbnail"
  src=""
/>
`;

exports[`should render the placeholder image 1`] = `
<div>
  placeholder text
</div>
`;

exports[`should render the src image 1`] = `
<img
  alt="profile"
  className="image fill default"
  height="75"
  onError={[Function]}
  onLoad={[Function]}
  src="profile.jpg"
  width="75"
/>
`;
